### ISSUE TITLE
Request for spatial softmax implementation

### ISSUE THREAD

**Issue by ahundt** [tensorflow/tensorflow#6271]
Spatial softmax is defined in [End-to-End Training of Deep Visuomotor Policies](https://arxiv.org/abs/1504.00702). This is a request for an implementation in the TensorFlow API, or if the dimension flag for the [regular softmax](https://www.tensorflow.org/versions/r0.12/api_docs/python/nn.html#softmax) does this, could the term and an example be explicitly added to the g3doc?

It seems this may already be implemented with tf but not yet incorporated upstream according to [Collective Robot Reinforcement Learning with Distributed Asynchronous Guided Policy Search](https://arxiv.org/pdf/1610.00673.pdf):

> This entire system, which we call asynchronous distributed
guided policy search (ADGPS), was implemented in the
distributed machine learning framework TensorFlow [...snip...] 
Our architecture resembles prior work, with the visual features 
represented by feature points produced via a spatial softmax
applied to the last convolutional
response maps.

**Comment by kalakris** [comment]
The term "spatial softmax" is a bit of a misnomer - it should have probably been called spatial soft-argmax, since it's function is to return the expected pixel locations of each feature map. The softmax with the dim flag is not enough in itself, but is a useful tool in implementing the spatial soft-argmax. This can be achieved using a few lines of code in stock TensorFlow. You can first compute the spatial softmax of the features over the image dimensions with a little reshaping and transposing of the input:

```python
# Assume features is of size [N, H, W, C] (batch_size, height, width, channels).
# Transpose it to [N, C, H, W], then reshape to [N * C, H * W] to compute softmax
# jointly over the image dimensions. 
features = tf.reshape(tf.transpose(features, [0, 3, 1, 2]), [N * C, H * W])
softmax = tf.nn.softmax(features)
# Reshape and transpose back to original format.
softmax = tf.transpose(tf.reshape(softmax, [N, C, H, W]), [0, 2, 3, 1])
```

The above is the spatial softmax, which can be used as weights to compute the mean pixel locations of each channel as follows:

```python
# Assume that image_coords is a tensor of size [H, W, 2] representing the image
# coordinates of each pixel.
# Convert softmax to shape [N, H, W, C, 1]
softmax = tf.expand_dims(softmax, -1)
# Convert image coords to shape [H, W, 1, 2]
image_coords = tf.expand_dims(image_coords, 2)
# Multiply (with broadcasting) and reduce over image dimensions to get the result
# of shape [N, C, 2]
spatial_soft_argmax = tf.reduce_sum(softmax * image_coords, reduction_indices=[1, 2])
```

Hope this helps.

**Comment by vrv** [comment]
Thanks Mrinal!  Closing since this is no longer a feature request.

**Comment by ahundt** [comment]
@kalakris Thanks Mrinal! To confirm, the channels are features at that (h,w) coordinate for each problem in the batch, correct?

**Comment by kalakris** [comment]
Yes that's correct - I used the terms 'channels' and 'features' interchangeably in my description above.

**Comment by ahundt** [comment]
[spatial_softmax](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L2619) is now directly available in `layers.py`.

**Comment by jorisguerin** [comment]
Hi everyone,

I have implemented the spatial softmax as well, in a similar fashion that proposed here.
It works well in the forward pass. However, whenever trying to train a network containing this layer, it appears that after a single pass, it outputs either the center point or one of the corners.

Does anyone have an idea of what might cause this behavior? Any possible fixes?

The function I use is the following:

```python
def spatial_softArgmax(filters, temperature = 1.0):
    
    shape = tf.shape(filters)
    height, width, num_channels = shape[1], shape[2], shape[3]
    
    posx, posy = tf.meshgrid(tf.lin_space(-1., 1., num = height), 
                             tf.lin_space(-1., 1., num = width))
    
    posx = tf.reshape(posx, [height * width])
    posy = tf.reshape(posy, [height * width])
    
    filters = tf.reshape(tf.transpose(filters, [0, 3, 1, 2]), [-1, height * width])
    
    softmax_attention = tf.nn.softmax(filters / temperature)
    
    expected_x = tf.reduce_sum(posx * softmax_attention, 1, keep_dims = True)
    expected_y = tf.reduce_sum(posy * softmax_attention, 1, keep_dims = True)
    
    expected_xy = tf.concat([expected_x, expected_y], axis = 1)
    
    feature_keypoints = tf.reshape(expected_xy, [-1, num_channels * 2])

    return feature_keypoints
```

**Comment by acrosson** [comment]
@jorisguerin did you figure it out?

**Comment by jorisguerin** [comment]
@acrosson Yes I did, first there was an error in my function, one should change the following line:
```python
posx, posy = tf.meshgrid(tf.lin_space(-1., 1., num = height), 
                             tf.lin_space(-1., 1., num = width))
```
to
```python
posx, posy = tf.meshgrid(tf.lin_space(-1., 1., num = height), 
                             tf.lin_space(-1., 1., num = width), 
                             indexing='ij')
```

Also, this behavior tends to happen for some reason. I found out two possible ways of avoiding it:

- Reduce your learning rate
- Add dropouts after your Spacial softArgMax layer

Hope this helps.

**Comment by ahundt** [comment]
FYI spatial softmax has been available directly in tensorflow since 1.3, see:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L2606

**Comment by bhack** [comment]
https://github.com/tensorflow/addons/issues/1364

**Comment by Sreerag-ibtl** [comment]
Thanks @kalakris That worked great. But an issue I am facing is, since this involves a product of N*C, I am unable to use dynamic batch size since None*int is forbidden. Is this impossible or any workaround you know about? TIA
